File: app/src/main/java/com/orlinskas/notebook/mVVM/fragment/DayView.java
package com.orlinskas.notebook.mVVM.fragment;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.ViewModelProviders;

import com.orlinskas.notebook.Enums;
import com.orlinskas.notebook.R;
import com.orlinskas.notebook.notificationHelper.NotificationListAdapter;
import com.orlinskas.notebook.mVVM.model.Day;

import java.util.Objects;

import static com.orlinskas.notebook.Constants.COUNT_NOTIFICATION_IN_SHORT_LIST;
import static com.orlinskas.notebook.Constants.DAY_ID;
import static com.orlinskas.notebook.Constants.IS_FULL_DISPLAY;

public class DayView extends Fragment {
    private int dayID;
    private DayFragmentActions fragmentActions;
    private Context context;
    private boolean isFullDisplay;
    private ProgressBar progressBar;
    private DayViewModel model;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        this.context = context;
        fragmentActions = (DayFragmentActions) context;
        model = ViewModelProviders.of(this).get(DayViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_day, container, false);

        progressBar = view.findViewById(R.id.fragment_day_pb);

        if(savedInstanceState != null) {
            dayID = savedInstanceState.getInt(DAY_ID);
        }
        else {
            if (getArguments() != null) {
                dayID = getArguments().getInt(DAY_ID);
                isFullDisplay = getArguments().getBoolean(IS_FULL_DISPLAY);
            }
        }

        model.getDaysData().observe(this, days ->
                updateUI(days.get(dayID), view));

        model.getRepositoryStatusData().observe(this, status -> {
            if (status == Enums.RepositoryStatus.LOADING) {
                progressBar.setVisibility(View.VISIBLE);
            }
            if (status == Enums.RepositoryStatus.READY) {
                progressBar.setVisibility(View.INVISIBLE);
            }
        });

        return view;
    }

    private void updateUI(Day day, View view) {
        TextView dayName = view.findViewById(R.id.fragment_notification_tv_day_name);
        TextView dayDate = view.findViewById(R.id.fragment_notification_tv_day_date);
        ListView notificationList = view.findViewById(R.id.fragment_notification_ll_notification_container);

        dayName.setText(day.getDayName());
        dayDate.setText(day.getDayDate());

        ArrayAdapter adapter;

        if(isFullDisplay){
            adapter = new NotificationListAdapter(context, R.layout.item_notification_full, day.getNotifications());
        }
        else {
            adapter = new NotificationListAdapter(context, R.layout.item_notification_small, day.getNotifications());
        }

        notificationList.setAdapter(adapter);
        setListViewHeightBasedOnChildren(notificationList);

        final Animation clickAnimation = AnimationUtils.loadAnimation(context, R.anim.animation_button);

        view.setOnClickListener(v -> {
            v.startAnimation(clickAnimation);
            fragmentActions.openDay(dayID);
        });
        notificationList.setOnItemClickListener((parent, view1, position, id) -> {
            view1.startAnimation(clickAnimation);
            fragmentActions.openDay(dayID);
        });

        if(isFullDisplay){
            notificationList.setOnItemLongClickListener((parent, view12, position, id) -> {
                FragmentManager manager = getFragmentManager();
                DeleteDialogFragment deleteDialogFragment
                        = new DeleteDialogFragment(day.getNotifications().get(position));
                deleteDialogFragment.show(Objects.requireNonNull(manager), "dialog");
                return true;
            });
        }
    }

    private void setListViewHeightBasedOnChildren(ListView listView) {
        ArrayAdapter listAdapter = (ArrayAdapter) listView.getAdapter();

        int totalHeight = 0;
        int maxCount;

        if(listAdapter.getCount() >= COUNT_NOTIFICATION_IN_SHORT_LIST && !isFullDisplay){
            maxCount = COUNT_NOTIFICATION_IN_SHORT_LIST;
        }
        else {
            maxCount = listAdapter.getCount();
        }

        for (int i = 0; i < maxCount; i++) {
            View listItem = listAdapter.getView(i, null, listView);
            listItem.measure(0, 0);
            totalHeight += listItem.getMeasuredHeight();
        }

        ViewGroup.LayoutParams params = listView.getLayoutParams();
        params.height = totalHeight + (listView.getDividerHeight() * (listAdapter.getCount() - 1));
        listView.setLayoutParams(params);
        listView.requestLayout();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(DAY_ID, dayID);
    }
}

File: app/src/main/java/com/orlinskas/notebook/mvvm/fragment/DayView.kt
package com.orlinskas.notebook.mvvm.fragment

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.ArrayAdapter
import android.widget.ListView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.orlinskas.notebook.Constants.*
import com.orlinskas.notebook.R
import com.orlinskas.notebook.mvvm.model.Day
import com.orlinskas.notebook.notificationHelper.NotificationListAdapter
import kotlinx.android.synthetic.main.fragment_day.*
import java.util.*

class DayView : Fragment() {
    private var dayID: Int = 0
    private var isFullDisplay: Boolean = false
    private lateinit var fragmentActions: DayFragmentActions
    private lateinit var model: DayViewModel

    override fun onAttach(context: Context) {
        super.onAttach(context)
        fragmentActions = context as DayFragmentActions
        model = ViewModelProviders.of(this).get(DayViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_day, container, false)

        if (savedInstanceState != null) {
            dayID = savedInstanceState.getInt(DAY_ID)
        } else {
            arguments?.let {
                dayID = it.getInt(DAY_ID)
                isFullDisplay = it.getBoolean(IS_FULL_DISPLAY)
            }
        }

        model.daysData.observe(this, Observer {
            updateUI(it[dayID], view)
        })

        return view
    }

    private fun updateUI(day: Day, view: View) {
        val notificationList = fragment_notification_ll_notification_container

        fragment_notification_tv_day_name.text = day.dayName
        fragment_notification_tv_day_date.text = day.dayDate

        val adapter: ArrayAdapter<*> = if (isFullDisplay) {
            NotificationListAdapter(context, R.layout.item_notification_full, day.notifications)
        } else {
            NotificationListAdapter(context, R.layout.item_notification_small, day.notifications)
        }

        notificationList.adapter = adapter
        setListViewHeightBasedOnChildren(notificationList)

        val clickAnimation = AnimationUtils.loadAnimation(context, R.anim.animation_button)

        view.setOnClickListener {
            it.startAnimation(clickAnimation)
            fragmentActions.openDay(dayID)
        }

        notificationList.setOnItemClickListener { _, _, _, _ ->
            notificationList.startAnimation(clickAnimation)
            fragmentActions.openDay(dayID)
        }

        if (isFullDisplay) {
            notificationList.setOnItemLongClickListener { _, _, position, _ ->
                val manager = fragmentManager
                val deleteDialogFragment = DeleteDialogFragment(day.notifications[position])
                deleteDialogFragment.show(Objects.requireNonNull<FragmentManager>(manager), "dialog")
                true
            }
        }
    }

    private fun setListViewHeightBasedOnChildren(listView: ListView) {
        val listAdapter = listView.adapter as ArrayAdapter<*>

        var totalHeight = 0
        val maxCount: Int

        maxCount = if (listAdapter.count >= COUNT_NOTIFICATION_IN_SHORT_LIST && !isFullDisplay) {
            COUNT_NOTIFICATION_IN_SHORT_LIST
        } else {
            listAdapter.count
        }

        for (i in 0 until maxCount) {
            val listItem = listAdapter.getView(i, null, listView)
            listItem.measure(0, 0)
            totalHeight += listItem.measuredHeight
        }

        val params = listView.layoutParams
        params.height = totalHeight + listView.dividerHeight * (listAdapter.count - 1)
        listView.layoutParams = params
        listView.requestLayout()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(DAY_ID, dayID)
    }
}